### DISCUSSION TITLE
How to implement "set black/gray/white point"?

### DISCUSSION THREAD

**Discussion by zanyc-smm** [colour-science/colour#1107] [Q&A]
Hi all,

Realizing this is a basic question, I'd very much appreciate a nudge in the right direction on how to implement the equivalent of "set black/gray/white point" in Photoshop or Gimp. It seems as though there's an elegant way to do it that I'm simply not seeing.

I'm in linear color space, and I have a swatch from a color checker that defines the white point that I want to enforce.

Thank you very much for any help!

**Comment by thomasmansencal** [comment]
Hi @zanyc-smm,

I would need to double-check how it is done in Photoshop but I can think about 3 ways:

- 3 points linear interpolation, `np.interp1d`, `colour.LinearInterpolator`
- 1D LUT with 3 points, `colour.LUT1D`
- Approximation using a linear conversion + gamma, `colour.algebra.linear_conversion`, `colour.models.gamma_function`

Let us know how it goes!

Cheers,

Thomas

**Comment by zanyc-smm** [comment]
Thank you, Thomas. After reading your suggestions and reconsidering my original question, I now understand that I've asked the wrong question.

Let me please rephrase with some additional detail.

After applying a color correction, and then neutralizing using swatch[-3] multiplied out to the D65 illuminant, I get the following:

![zanyc-cc-only](https://user-images.githubusercontent.com/122826676/219899514-bb182589-a37d-4270-807f-03f4c169bb20.jpg)

^ Where the left is the resulting color-corrected image, and the right is the original.

In the following, I've manually gone into Gimp and matched the luminosity of the swatch[-6] (white, lower left) in the color-corrected result to that of the original image:

![zanyc-cc-exp-match](https://user-images.githubusercontent.com/122826676/219899615-211fe86e-a3aa-4ed1-9216-553aca2de522.jpg)

^ Where the left is now my manual exposure correction, and the right remains the original.

This is the goal that I'm trying to achieve. I realize that there are many factors and levels of approximation (and accuracy) here and that using a spectrometer and also testing the reflectivity of the color checker would yield the most accurate result ... however, any guidance that you might be able to provide on how to achieve an automated approximation would be most helpful.

Several discussions are circling this topic, e.g., inferring illumination, using spectral methods, and my own layperson's manual exposure matching. I'm too new to the community to understand their individual nuances. I would appreciate any detail to point me in the right direction to keep reading and moving the needle!

Thank you again!

**Reply by thomasmansencal** [reply]
Hi @zanyc-smm,

There are a few things to consider and unpack:

# Illumination

Typically, and for successful calibration, you will want your lighting to be as uniform as possible:

- Minimum vignette from the lens and lighting
- No shadows
- 45:0 measurement geometry: Imaging instrument is normal to the ColorChecker and lighting is at 45deg compared to it.

Here is such an image:

![image](https://user-images.githubusercontent.com/99779/219901361-6726af50-e624-4705-9a14-f03a38f51625.png)

Your lighting does not have those characteristics here.

# Processing

The image should be in a Scene-Referred state, i.e. devoid of any processing that makes it faithful to the original scene to the viewer and pleasing on display. It should also and ideally be in encoded linearly or with a known log curve one like the image I attached. I'm tempted to say that your image is currently processed for display and thus is not in an appropriate state to do calibration.

# Other Critical Considerations

We have written a blog post that is relevant here: https://www.colour-science.org/posts/the-colorchecker-considered-mostly-harmless/

Cheers,

Thomas

**Comment by zanyc-smm** [comment]
Thank you, Thomas. Based on your insights, I was able to get this across the line.
